api/v1alpha1/stage_types.go

```diff
@@ -163,23 +164,24 @@ type StageSpec struct {
 	Verification *Verification `json:"verification,omitempty" protobuf:"bytes,3,opt,name=verification"`
 }
 
-// Subscriptions describes a Stage's sources of Freight.
-type Subscriptions struct {
-	// Warehouse is a subscription to a Warehouse. This field is mutually
-	// exclusive with the UpstreamStages field.
-	Warehouse string `json:"warehouse,omitempty" protobuf:"bytes,1,opt,name=warehouse"`
-	// UpstreamStages identifies other Stages as potential sources of Freight
-	// for this Stage. This field is mutually exclusive with the Repos field.
-	UpstreamStages []StageSubscription `json:"upstreamStages,omitempty" protobuf:"bytes,2,rep,name=upstreamStages"`
-}
-
-// StageSubscription defines a subscription to Freight from another Stage.
-type StageSubscription struct {
-	// Name specifies the name of a Stage.
+// FreightSources describes the sources (e.g. Warehouses or upstream Stages)
+// from which to obtain the various types of Freight a Stage requires.
+type FreightSources struct {
+	// Type specifies the type of Freight to obtain (i.e. specifies what Warehouse
+	// the Freight must have originated from). This field is required.
 	//
-	// +kubebuilder:validation:MinLength=1
-	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
-	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// +kubebuilder:validation:Required
+	Type string `json:"type" protobuf:"bytes,3,opt,name=type"`
+	// WarehouseDirect indicates whether Freight of the desired type may be
+	// obtained directly from its Warehouse of origin. If this field's value is
+	// false, then the value of the UpstreamStages field must be non-empty.
+	// i.e. Between the two fields, at least on source must be specified.
+	WarehouseDirect bool `json:"warehouseDirect,omitempty" protobuf:"varint,1,opt,name=warehouseDirect"`
```

Comment:
```suggestion
	// Origin specifies the origin of Freight to obtain (i.e. specifies what Warehouse
	// the Freight must have originated from). This field is required.
	//
	// +kubebuilder:validation:Required
	Origin string `json:"origin" protobuf:"bytes,3,opt,name=origin"`

	// Direct indicates whether Freight of the desired origin may be
	// obtained directly from its Warehouse of origin. If this field's value is
	// false, then the value of the UpstreamStages field must be non-empty.
	// i.e. Between the two fields, at least one source must be specified.
	Direct bool `json:"direct,omitempty" protobuf:"varint,1,opt,name=direct"`
```

I still feel something like this would be less ambiguous, as "type" could also refer to a "source type". Another, more explicit and descriptive field name, could potentially be `Warehouse`.

Reply:
Please feel free to change it. I don't consider _anything_ in this PR to be set in stone.